tiatoolbox/tools/patchextraction.py

```diff
@@ -0,0 +1,330 @@
+from tiatoolbox.utils.misc import read_point_annotations
+
+# import math
+# import pathlib
+
+
+class PatchExtractor(ABC):
+    """
+    Class for extracting and merging patches in standard and whole-slide images.
+
+    Args:
+        input_img(str, pathlib.Path, ndarray): input image for patch extraction.
+        patch_size(Tuple of int): patch size tuple (width, height).
+        pad_size(Tuple of int): symmetric padding in (x, y) direction.
+
+    Attributes:
+        input_img(ndarray, WSIReader): input image for patch extraction.
+          input_image type is ndarray for an image tile whereas :obj:`WSIReader`
+          for an WSI.
+        patch_size(Tuple of int): patch size tuple (width, height).
+        pad_size(Tuple of int): symmetric padding in (x, y) direction.
+        n(int): current state of the iterator.
+
+    """
+
+    def __init__(self, input_img, patch_size, pad_size=0):
+        self.patch_size = patch_size
+        self.pad_size = pad_size
+        self.n = 0
+        self.wsi = get_wsireader(input_img=input_img)
+
+    def __iter__(self):
+        self.n = 0
+        return self
+
+    def __next__(self):
+        raise NotImplementedError
+
+    def __getitem__(self, item):
+        raise NotImplementedError
+
+    # @staticmethod
+    # def __get_last_steps(image_dim, label_patch_dim, stride):
+    #     """Get the last location for patch extraction in a specific direction.
+    #
+    #     Args:
+    #         image_dim: 1D size of image
+    #         label_patch_dim: 1D size of patches
+    #         stride: 1D size of stride for patch extraction
+    #
+    #     Returns:
+    #         last_step: the final location for patch extraction
+    #     """
+    #     nr_step = math.ceil((image_dim - label_patch_dim) / stride)
+    #     last_step = (nr_step + 1) * stride
+    #     return int(last_step)
+
+    # def extract_patches(self):
+    #     """Extract patches from an image using locations provided by labels data.
+    #
+    #     Args:
+    #         labels (str, ndarray):
+    #
+    #     Returns:
+    #         img_patches (ndarray): extracted image patches of size NxHxWxD.
+    #     """
+    #     raise NotImplementedError
+
+    def merge_patches(self, patches):
+        """Merge the patch-level results to get the overall image-level prediction.
+
+        Args:
+            patches: patch-level predictions
+
+        Returns:
+            image: merged prediction
+        """
+        raise NotImplementedError
+
+
+class FixedWindowPatchExtractor(PatchExtractor):
+    """Extract and merge patches using fixed sized windows for images and labels.
+
+    Args:
+        stride(Tuple of int): stride in (x, y) direction for patch extraction.
+
+    Attributes:
+        stride(Tuple of int): stride in (x, y) direction for patch extraction.
+
+    """
+
+    def __init__(
+        self,
+        input_img,
+        patch_size,
+        pad_size=0,
```

Comment:
Where this `pad_size` is being used? if the goal is to pad image to extract patches beyond image size, I think the new PR #78 would resolve that and we do not need to have this argument here. Unless, we want user to decide have padded patches or valid sized patches in which case this should be binary flag.

Reply:
We may not need it as we can now read beyond boundaries but this PR is focused on points patch extraction and setting up the code for other patch extraction methods.